pkg/k8s/service_cache.go

```diff
@@ -291,17 +295,25 @@ func (s *ServiceCache) UpdateEndpoints(newEndpoints *Endpoints, swg *lock.Stoppa
 
 	eps.Upsert(esID.EndpointSliceName, newEndpoints)
 
+	oldEPs, _ := cache[key]
+	// cache is optional as it's only needed for Agents (not for Operator). If
+	// it's nil, then we don't expect to use oldEPs anyway.
+	if cache != nil {
+		cache[key] = newEndpoints
+	}
```

Comment:
Why is this external cache necessary? I think it should be possible to retrieve the oldEPs from `eps.epSlices[esID.EndpointSliceName]`, which is already kept in sync.